src/plugins/explore/public/components/visualizations/utils/data_transformation.ts

```diff
@@ -185,40 +184,244 @@ export const aggregateByTime = (
+  aggregationType,
+  groupField,
+}: {
+  aggregationType?: AggregationType | undefined;
+  groupField?: VisColumn;
+}): PipelineFn<T> => (state) => {
+  const { data, axisConfig, axisColumnMappings } = state;
+  if (!axisConfig) {
+    throw new Error('axisConfig must be derived before prepareData');
+  }
+
+  const categoricalColumn = [axisConfig.xAxis, axisConfig.yAxis].find(
+    (axis) => axis?.schema === VisFieldType.Categorical
+  );
+  const numericalColumn = [axisConfig.xAxis, axisConfig.yAxis].find(
+    (axis) => axis?.schema === VisFieldType.Numerical
+  );
+
+  const colorColumn = axisColumnMappings?.color?.column;
+  // use categoricalColumn as default
+  const groupByMe = groupField ? groupField.column : categoricalColumn?.column;
+  const valueField = numericalColumn?.column;
+
+  if (colorColumn && groupByMe && valueField) {
+    const categorical2Collection = Array.from(new Set(data.map((item) => item[colorColumn])));
+
+    const grouped = data.reduce((acc, row) => {
+      const group = String(row[groupByMe]);
+      const color = String(row[colorColumn]);
+      const value = Number(row[valueField]);
+
+      if (isNaN(value)) return acc;
+
+      acc[group] ??= {};
+      acc[group][color] ??= [];
+      acc[group][color].push(value);
+
+      return acc;
+    }, {} as Record<string, Record<string, number[]>>);
+
+    if (aggregationType) {
+      // Aggregate data with same second_group. used in bar chart
+      const aggregatedRes = Object.entries(grouped).map(([group, colorValues]) => {
+        return [
+          group,
+          ...categorical2Collection.map(
+            (color) => aggregateValues(aggregationType, colorValues[color]) ?? null
+          ),
+        ];
+      });
+
+      return {
+        ...state,
+        aggregatedData: [
+          [groupByMe, ...categorical2Collection.map((c) => String(c))],
+          ...aggregatedRes,
+        ],
+        categorical2Collection,
+      };
+    } else {
+      // simply pivot data without aggregation, used in multi-lines and stack-area
+      const pivotedRes = [];
+      for (const [group, colorValues] of Object.entries(grouped)) {
+        const maxLength = Math.max(
+          ...categorical2Collection.map((color) => colorValues[color]?.length || 0)
+        );
+        for (let i = 0; i < maxLength; i++) {
+          pivotedRes.push([
+            group,
+            ...categorical2Collection.map((color) => colorValues[color]?.[i] ?? null),
+          ]);
         }
       }
 
-      return [date, aggregatedValue];
-    })
-    .sort((a, b) => a[0].getTime() - b[0].getTime()); // Sort by time
+      const sorted =
+        groupField?.schema === VisFieldType.Date
+          ? [...pivotedRes].sort((a, b) => new Date(a[0]).getTime() - new Date(b[0]).getTime()) // Sort by time;;,
+          : pivotedRes;
+
+      return {
+        ...state,
+        aggregatedData: [[groupByMe, ...categorical2Collection.map((c) => String(c))], ...sorted],
```

Comment:
Seems don't actually need to return `aggregatedData` right? Is it used anywhere?